source/extensions/filters/http/oauth2/filter.h

```diff
@@ -0,0 +1,232 @@
+  FilterStats& stats() { return stats_; }
+
+private:
+  static FilterStats generateStats(const std::string& prefix, Stats::Scope& scope);
+
+  const envoy::config::core::v3::HttpUri oauth_token_endpoint_;
+  const std::string authorization_endpoint_;
+  const std::string client_id_;
+  const std::string redirect_uri_;
+  const Matchers::PathMatcher redirect_matcher_;
+  const Matchers::PathMatcher signout_path_;
+  std::shared_ptr<SecretReader> secret_reader_;
+  FilterStats stats_;
+  const bool forward_bearer_token_ : 1;
+  const std::vector<Http::HeaderUtility::HeaderData> pass_through_header_matchers_;
+};
+
+using FilterConfigSharedPtr = std::shared_ptr<FilterConfig>;
+
+/**
+ * An OAuth cookie validator:
+ * 1. extracts cookies from a request
+ * 2. HMAC/encodes the values
+ * 3. Compares the result to the cookie HMAC
+ * 4. Checks that the `expires` value is valid relative to current time
+ *
+ * Required components:
+ * - header map
+ * - secret
+ */
+class CookieValidator {
+public:
+  virtual ~CookieValidator() = default;
+  virtual const std::string& token() const PURE;
+  virtual void setParams(const Http::RequestHeaderMap& headers, const std::string& secret) PURE;
+  virtual bool isValid() const PURE;
+};
+
+class OAuth2CookieValidator : public CookieValidator {
+public:
+  explicit OAuth2CookieValidator(TimeSource& time_source) : time_source_(time_source) {}
+
+  const std::string& token() const override { return token_; }
+  void setParams(const Http::RequestHeaderMap& headers, const std::string& secret) override;
+  bool isValid() const override;
+  bool hmacIsValid() const;
+  bool timestampIsValid() const;
+
+private:
+  std::string token_;
+  std::string expires_;
+  std::string hmac_;
+  std::vector<uint8_t> secret_;
+  absl::string_view host_;
+  TimeSource& time_source_;
+};
+
+/**
+ * The filter is the primary entry point for the OAuth workflow. Its responsibilities are to
+ * receive incoming requests and decide at what state of the OAuth workflow they are in. Logic
+ * beyond that is broken into component classes.
+ */
+class OAuth2Filter : public Http::PassThroughDecoderFilter, public FilterCallbacks {
+public:
+  OAuth2Filter(FilterConfigSharedPtr config, std::unique_ptr<OAuth2Client>&& oauth_client,
+               TimeSource& time_source);
+
+  // Http::PassThroughDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers, bool) override;
+
+  // FilterCallbacks
+  void onGetAccessTokenSuccess(const std::string& access_code,
+                               std::chrono::seconds expires_in) override;
+  // a catch-all function used for request failures. we don't retry, as a user can simply refresh
+  // the page in the case of a network blip.
+  void sendUnauthorizedResponse() override;
+
+  // Sets the auth token as the Bearer token in the authorization header.
+  static void setBearerToken(Http::RequestHeaderMap& headers, const std::string& token);
```

Comment:
This can be in `.cc` file unnamed namespace? I don't think you need an explicit test for this function.